[dynamo] Add LazyVariableTracker

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112961
* #112962
* #111726
* #111725
* #111415
* __->__ #111306



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

jansel committed:
[dynamo] Add LazyVariableTracker

[ghstack-poisoned]

Chillee commented:
Is this for improving trace times?

jansel commented:
The idea would be to make all guards eagerly installed on `tx.ouput`, and remove the per-VariableTracker `.guards`.  There have been a bunch of bugs related to that.

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -436,9 +437,10 @@ def index_source(key):
-                k: VariableBuilder(
-                    self.tx, GetItemSource(self.get_source(), index_source(k))
-                )(value[k]).add_guards(guards)
+                k: LazyVariableTracker(
```
Feels like export should always just realize here too?

jansel replied:
I think the one in symbolic_convert.py should cover it.

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented:
Fixed model failures, added comments, todos, minor stuff. Looking at eager guards now.

voznesenskym commented on torch/_dynamo/variables/lazy.py:
```diff
@@ -0,0 +1,108 @@
+
+    def realize(self) -> VariableTracker:
+        """Force construction of the real VariableTracker"""
+        if self.mutable_local.vt is None:
```
`self.mutable_local.vt` feels ripe for a @property ;)

voznesenskym reviewed (approved):
![image](https://github.com/pytorch/pytorch/assets/4755252/b04ca485-1e7c-4b40-ab72-ade90b98111d)

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/lazy.py:
```diff
@@ -0,0 +1,108 @@
+        self._value = _value
+        if self.mutable_local is None:
+            self.mutable_local = LazyMutableLocal()
+        # TODO(voz) - Support SourcelessBuilder during realize
```
No point in doing this, since SourcelessBuilder can't create guards

voznesenskym replied:
Fair point. I was thinking there might be value for LazyVariableTracker for intermediary types, but, thinking more, its not that useful.

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -121,19 +121,6 @@ def as_proxy(self):
     def python_type(self):
         return self.class_type
 
-    def call_isinstance(self, tensor_type):
-        def check_type(ty):
-            if ty not in tensortype_to_dtype:
-                return issubclass(self.python_type(), ty)
-
-            dtypes = tensortype_to_dtype[ty]
-            return self.dtype in dtypes
-
-        if type(tensor_type) is tuple:
-            return any(check_type(ty) for ty in tensor_type)
-        else:
-            return check_type(tensor_type)
-
```
@voz did you remove this?  If so why?

voznesenskym replied:
It's moved into `torch/_dynamo/variables/builtin.py` - this is the only place that has this pattern, and it was causing a bug, where we were trying to invoke this incorrectly.

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/builder.py:
```diff
@@ -443,9 +444,10 @@ def index_source(key):
-                k: VariableBuilder(
-                    self.tx, GetItemSource(self.get_source(), index_source(k))
-                )(value[k]).add_guards(guards)
+                k: LazyVariableTracker.create(
+                    value[k],
+                    source=GetItemSource(self.get_source(), index_source(k)),
+                ).add_guards(guards)
```
How come this is used in particular for values of dictionaries?

jansel replied:
The next PR in the stack turns it on in more places.  Dict keys seem like something more likely to be unused.

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50979121).

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D50979121](https://our.internmc.facebook.com/intern/diff/D50979121)

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Add LazyVariableTracker"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
